File: chapters/chp2_code/parametric_curves.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <array>
#include <iomanip>

struct Vector3D {
    double x, y, z;
    
    Vector3D(double x = 0, double y = 0, double z = 0) : x(x), y(y), z(z) {}
    
    Vector3D operator+(const Vector3D& other) const {
        return Vector3D(x + other.x, y + other.y, z + other.z);
    }
    
    Vector3D operator-(const Vector3D& other) const {
        return Vector3D(x - other.x, y - other.y, z - other.z);
    }
    
    Vector3D operator*(double scalar) const {
        return Vector3D(x * scalar, y * scalar, z * scalar);
    }
    
    double magnitude() const {
        return std::sqrt(x*x + y*y + z*z);
    }
    
    void print() const {
        std::cout << "(" << x << ", " << y << ", " << z << ")";
    }
};

class ParametricCurve {
protected:
    double dt = 0.001;  // Small increment for numerical differentiation
    
public:
    virtual Vector3D position(double t) const = 0;
    
    Vector3D velocity(double t) const {
        Vector3D r_plus = position(t + dt);
        Vector3D r_minus = position(t - dt);
        return (r_plus - r_minus) * (1.0 / (2 * dt));
    }
    
    Vector3D acceleration(double t) const {
        Vector3D v_plus = velocity(t + dt);
        Vector3D v_minus = velocity(t - dt);
        return (v_plus - v_minus) * (1.0 / (2 * dt));
    }
    
    double speed(double t) const {
        return velocity(t).magnitude();
    }
    
    double arc_length(double t_start, double t_end, int num_points = 1000) const {
        double length = 0;
        double step = (t_end - t_start) / num_points;
        
        for (int i = 0; i < num_points; ++i) {
            double t1 = t_start + i * step;
            double t2 = t_start + (i + 1) * step;
            Vector3D r1 = position(t1);
            Vector3D r2 = position(t2);
            length += (r2 - r1).magnitude();
        }
        
        return length;
    }
    
    void generate_curve_data(double t_start, double t_end, int num_points, 
                            const std::string& filename) const {
        std::ofstream file(filename);
        file << std::fixed << std::setprecision(6);
        file << "t,x,y,z,vx,vy,vz,speed\n";
        
        double step = (t_end - t_start) / (num_points - 1);
        
        for (int i = 0; i < num_points; ++i) {
            double t = t_start + i * step;
            Vector3D r = position(t);
            Vector3D v = velocity(t);
            double s = v.magnitude();
            
            file << t << "," << r.x << "," << r.y << "," << r.z << ","
                 << v.x << "," << v.y << "," << v.z << "," << s << "\n";
        }
        
        file.close();
        std::cout << "Curve data written to " << filename << std::endl;
    }
};

class Helix : public ParametricCurve {
private:
    double a, b;  // a = radius, b = vertical speed
    
public:
    Helix(double radius = 1.0, double vertical_speed = 0.5) 
        : a(radius), b(vertical_speed) {}
    
    Vector3D position(double t) const override {
        return Vector3D(a * std::cos(t), a * std::sin(t), b * t);
    }
};

class Lissajous : public ParametricCurve {
private:
    double A, B, a, b, delta;
    
public:
    Lissajous(double A = 1, double B = 1, double a = 3, double b = 2, double delta = M_PI/2)
        : A(A), B(B), a(a), b(b), delta(delta) {}
    
    Vector3D position(double t) const override {
        return Vector3D(A * std::sin(a * t + delta), B * std::sin(b * t), 0);
    }
};

class TorusKnot : public ParametricCurve {
private:
    double R, r, p, q;
    
public:
    TorusKnot(double R = 2, double r = 1, double p = 3, double q = 2)
        : R(R), r(r), p(p), q(q) {}
    
    Vector3D position(double t) const override {
        double x = (R + r * std::cos(q * t)) * std::cos(p * t);
        double y = (R + r * std::cos(q * t)) * std::sin(p * t);
        double z = r * std::sin(q * t);
        return Vector3D(x, y, z);
    }
};

int main() {
    std::cout << std::fixed << std::setprecision(3);
    std::cout << "Parametric Curves Analysis\n";
    std::cout << "===========================\n\n";
    
    // Example 1: Helix
    std::cout << "Example 1: Helix\n";
    Helix helix(1.0, 0.5);
    
    double t_point = M_PI;
    Vector3D r = helix.position(t_point);
    Vector3D v = helix.velocity(t_point);
    Vector3D a = helix.acceleration(t_point);
    double arc_len = helix.arc_length(0, t_point);
    
    std::cout << "At t = π:\n";
    std::cout << "Position: "; r.print(); std::cout << std::endl;
    std::cout << "Velocity: "; v.print(); std::cout << std::endl;
    std::cout << "Speed: " << v.magnitude() << std::endl;
    std::cout << "Acceleration: "; a.print(); std::cout << std::endl;
    std::cout << "Arc length from 0 to π: " << arc_len << std::endl;
    
    helix.generate_curve_data(0, 4*M_PI, 200, "helix.csv");
    
    // Example 2: Lissajous curve
    std::cout << "\n\nExample 2: Lissajous Curve\n";
    Lissajous lissajous(1, 1, 3, 2, M_PI/2);
    
    t_point = M_PI/4;
    r = lissajous.position(t_point);
    v = lissajous.velocity(t_point);
    
    std::cout << "At t = π/4:\n";
    std::cout << "Position: "; r.print(); std::cout << std::endl;
    std::cout << "Velocity: "; v.print(); std::cout << std::endl;
    std::cout << "Speed: " << v.magnitude() << std::endl;
    
    lissajous.generate_curve_data(0, 2*M_PI, 500, "lissajous.csv");
    
    // Example 3: Torus knot
    std::cout << "\n\nExample 3: Torus Knot (3,2)\n";
    TorusKnot knot(2, 1, 3, 2);
    
    t_point = 0;
    r = knot.position(t_point);
    v = knot.velocity(t_point);
    
    std::cout << "At t = 0:\n";
    std::cout << "Position: "; r.print(); std::cout << std::endl;
    std::cout << "Velocity: "; v.print(); std::cout << std::endl;
    std::cout << "Speed: " << v.magnitude() << std::endl;
    
    double total_length = knot.arc_length(0, 2*M_PI);
    std::cout << "Total arc length (one period): " << total_length << std::endl;
    
    knot.generate_curve_data(0, 2*M_PI, 1000, "torus_knot.csv");
    
    // Curvature analysis for helix
    std::cout << "\n\nCurvature Analysis for Helix:\n";
    std::cout << "t\tκ(t)\n";
    for (double t = 0; t <= 2*M_PI; t += M_PI/4) {
        v = helix.velocity(t);
        a = helix.acceleration(t);
        
        // Curvature formula: |v × a| / |v|³
        Vector3D cross(
            v.y * a.z - v.z * a.y,
            v.z * a.x - v.x * a.z,
            v.x * a.y - v.y * a.x
        );
        
        double curvature = cross.magnitude() / std::pow(v.magnitude(), 3);
        std::cout << t << "\t" << curvature << std::endl;
    }
    
    std::cout << "\nNote: For a helix with radius a and pitch b, ";
    std::cout << "the curvature is constant: κ = a/(a² + b²)\n";
    double theoretical_curvature = 1.0 / (1.0 + 0.5*0.5);
    std::cout << "Theoretical curvature: " << theoretical_curvature << std::endl;
    
    return 0;
}

File: chapters/chp2_code/parametric_curves.rb
#!/usr/bin/env ruby

class Vector3D
  attr_accessor :x, :y, :z
  
  def initialize(x, y, z)
    @x, @y, @z = x.to_f, y.to_f, z.to_f
  end
  
  def magnitude
    Math.sqrt(@x**2 + @y**2 + @z**2)
  end
  
  def -(other)
    Vector3D.new(@x - other.x, @y - other.y, @z - other.z)
  end
  
  def +(other)
    Vector3D.new(@x + other.x, @y + other.y, @z + other.z)
  end
  
  def *(scalar)
    Vector3D.new(@x * scalar, @y * scalar, @z * scalar)
  end
  
  def cross(other)
    Vector3D.new(
      @y * other.z - @z * other.y,
      @z * other.x - @x * other.z,
      @x * other.y - @y * other.x
    )
  end
  
  def to_s
    "(#{@x.round(3)}, #{@y.round(3)}, #{@z.round(3)})"
  end
end

class ParametricCurve
  def initialize(&block)
    @curve = block
    @dt = 0.001
  end
  
  def position(t)
    @curve.call(t)
  end
  
  def velocity(t)
    r_plus = position(t + @dt)
    r_minus = position(t - @dt)
    (r_plus - r_minus) * (1.0 / (2 * @dt))
  end
  
  def acceleration(t)
    v_plus = velocity(t + @dt)
    v_minus = velocity(t - @dt)
    (v_plus - v_minus) * (1.0 / (2 * @dt))
  end
  
  def speed(t)
    velocity(t).magnitude
  end
  
  def curvature(t)
    v = velocity(t)
    a = acceleration(t)
    
    cross_product = v.cross(a)
    cross_magnitude = cross_product.magnitude
    v_magnitude = v.magnitude
    
    return 0 if v_magnitude < 1e-10
    cross_magnitude / (v_magnitude**3)
  end
  
  def arc_length(t_start, t_end, num_points = 1000)
    length = 0
    step = (t_end - t_start) / num_points.to_f
    
    num_points.times do |i|
      t1 = t_start + i * step
      t2 = t_start + (i + 1) * step
      r1 = position(t1)
      r2 = position(t2)
      length += (r2 - r1).magnitude
    end
    
    length
  end
  
  def save_curve_data(t_start, t_end, num_points, filename)
    File.open(filename, 'w') do |file|
      file.puts "t,x,y,z,vx,vy,vz,speed,curvature"
      
      step = (t_end - t_start) / (num_points - 1).to_f
      
      num_points.times do |i|
        t = t_start + i * step
        r = position(t)
        v = velocity(t)
        k = curvature(t)
        
        file.puts "#{t},#{r.x},#{r.y},#{r.z},#{v.x},#{v.y},#{v.z},#{v.magnitude},#{k}"
      end
    end
    
    puts "Curve data written to #{filename}"
  end
  
  def analyze_properties(t_values)
    puts "Parameter\tPosition\t\tVelocity\t\tSpeed\t\tCurvature"
    puts "-" * 80
    
    t_values.each do |t|
      r = position(t)
      v = velocity(t)
      s = speed(t)
      k = curvature(t)
      
      puts "t=#{t.round(3)}\t\t#{r}\t#{v}\t#{s.round(3)}\t#{k.round(6)}"
    end
  end
end

# Define parametric curves
helix = ParametricCurve.new do |t|
  a, b = 1.0, 0.5
  Vector3D.new(a * Math.cos(t), a * Math.sin(t), b * t)
end

lissajous = ParametricCurve.new do |t|
  a, b, A, B, delta = 3, 2, 1, 1, Math::PI/2
  Vector3D.new(A * Math.sin(a * t + delta), B * Math.sin(b * t), 0)
end

cycloid = ParametricCurve.new do |t|
  R = 1
  Vector3D.new(R * (t - Math.sin(t)), R * (1 - Math.cos(t)), 0)
end

torus_knot = ParametricCurve.new do |t|
  R, r, p, q = 2, 1, 3, 2
  x = (R + r * Math.cos(q * t)) * Math.cos(p * t)
  y = (R + r * Math.cos(q * t)) * Math.sin(p * t)
  z = r * Math.sin(q * t)
  Vector3D.new(x, y, z)
end

viviani_curve = ParametricCurve.new do |t|
  a = 1
  Vector3D.new(a * (1 + Math.cos(t)), a * Math.sin(t), 2 * a * Math.sin(t/2))
end

puts "Parametric Curves Analysis (Ruby version)"
puts "=========================================="
puts

# Example 1: Helix
puts "Example 1: Helix"
puts "================"
t_point = Math::PI

r = helix.position(t_point)
v = helix.velocity(t_point)
a = helix.acceleration(t_point)
arc_len = helix.arc_length(0, t_point)
k = helix.curvature(t_point)

puts "At t = π:"
puts "Position: #{r}"
puts "Velocity: #{v}"
puts "Speed: #{v.magnitude.round(3)}"
puts "Acceleration: #{a}"
puts "Arc length from 0 to π: #{arc_len.round(3)}"
puts "Curvature: #{k.round(6)}"

helix.save_curve_data(0, 4*Math::PI, 200, "helix_ruby.csv")

# Theoretical verification for helix
a_helix, b_helix = 1.0, 0.5
theoretical_curvature = a_helix / (a_helix**2 + b_helix**2)
puts "Theoretical curvature for helix: #{theoretical_curvature.round(6)}"

# Example 2: Lissajous curve
puts "\n" + "="*50
puts "Example 2: Lissajous Curve (3:2 ratio)"
puts "======================================"

t_point = Math::PI/4
r_liss = lissajous.position(t_point)
v_liss = lissajous.velocity(t_point)

puts "At t = π/4:"
puts "Position: #{r_liss}"
puts "Velocity: #{v_liss}"
puts "Speed: #{v_liss.magnitude.round(3)}"

lissajous.save_curve_data(0, 2*Math::PI, 500, "lissajous_ruby.csv")

# Example 3: Cycloid
puts "\n" + "="*50
puts "Example 3: Cycloid"
puts "=================="

t_point = Math::PI
r_cyc = cycloid.position(t_point)
v_cyc = cycloid.velocity(t_point)
cycloid_arc = cycloid.arc_length(0, 2*Math::PI)

puts "At t = π:"
puts "Position: #{r_cyc}"
puts "Velocity: #{v_cyc}"
puts "Speed: #{v_cyc.magnitude.round(3)}"
puts "Arc length for one arch: #{cycloid_arc.round(3)}"
puts "Theoretical arc length: #{8} (8R for R=1)"

cycloid.save_curve_data(0, 4*Math::PI, 1000, "cycloid_ruby.csv")

# Example 4: Torus Knot
puts "\n" + "="*50
puts "Example 4: Torus Knot (3,2)"
puts "==========================="

t_point = 0
r_knot = torus_knot.position(t_point)
v_knot = torus_knot.velocity(t_point)
knot_length = torus_knot.arc_length(0, 2*Math::PI)

puts "At t = 0:"
puts "Position: #{r_knot}"
puts "Velocity: #{v_knot}"
puts "Speed: #{v_knot.magnitude.round(3)}"
puts "Total arc length: #{knot_length.round(3)}"

torus_knot.save_curve_data(0, 2*Math::PI, 1000, "torus_knot_ruby.csv")

# Example 5: Viviani's curve
puts "\n" + "="*50
puts "Example 5: Viviani's Curve"
puts "=========================="

viviani_curve.save_curve_data(0, 4*Math::PI, 500, "viviani_ruby.csv")

# Curvature analysis
puts "\n" + "="*50
puts "Curvature Analysis Comparison"
puts "============================="

test_points = [0, Math::PI/4, Math::PI/2, 3*Math::PI/4, Math::PI]

puts "\nHelix Curvature Analysis:"
helix.analyze_properties(test_points)

puts "\nLissajous Curvature Analysis:"
lissajous.analyze_properties(test_points)

# Special properties and comparisons
puts "\n" + "="*50
puts "Special Properties"
puts "=================="

curves = {
  "Helix" => helix,
  "Lissajous" => lissajous,
  "Cycloid" => cycloid,
  "Torus Knot" => torus_knot,
  "Viviani" => viviani_curve
}

curves.each do |name, curve|
  puts "\n#{name}:"
  length = curve.arc_length(0, 2*Math::PI, 500)
  max_curvature = 0
  max_speed = 0
  
  20.times do |i|
    t = i * 2 * Math::PI / 19
    k = curve.curvature(t)
    s = curve.speed(t)
    max_curvature = k if k > max_curvature
    max_speed = s if s > max_speed
  end
  
  puts "  Arc length (0 to 2π): #{length.round(3)}"
  puts "  Maximum curvature: #{max_curvature.round(6)}"
  puts "  Maximum speed: #{max_speed.round(3)}"
end

# Mathematical insights
puts "\n" + "="*50
puts "Mathematical Insights"
puts "===================="

puts "\n1. Helix Properties:"
puts "   - Constant curvature: #{theoretical_curvature.round(6)}"
puts "   - Constant torsion (for 3D helices)"
puts "   - Geodesic on cylinders"

puts "\n2. Cycloid Properties:"
puts "   - Brachistochrone curve (fastest descent)"
puts "   - Tautochrone curve (equal time property)"
puts "   - Generated by rolling circle"

puts "\n3. Lissajous Properties:"
puts "   - Closed curve when frequency ratio is rational"
puts "   - Phase difference affects shape"
puts "   - Applications in oscilloscopes"

puts "\n4. Torus Knot Properties:"
puts "   - Lies on torus surface"
puts "   - Knot invariants depend on (p,q) parameters"
puts "   - Non-trivial topology"

puts "\nNote: All CSV files contain detailed parameter data for visualization."